wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/importer/BaseProjectImportOutputLineConsumer.java

```diff
@@ -0,0 +1,83 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.project.server.importer;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import org.eclipse.che.api.core.util.LineConsumer;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+
+import java.io.IOException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Base abstraction for consuming project importing output events.
+ * <p>
+ * Consumes output lines into delayed queue and perform broadcasting the last ones through each type of implementation.
+ * There are only two implementation of broadcasting type represented by {@link ProjectImportOutputWSLineConsumer} which
+ * broadcasts events through the web socket and {@link ProjectImportOutputJsonRpcLineConsumer} which broadcasts events
+ * through the json rpc protocol.
+ *
+ * @author Vlad Zhukovskyi
+ * @since 5.9.0
+ */
+public abstract class BaseProjectImportOutputLineConsumer implements LineConsumer {
+
+    protected final String                   projectName;
+    protected final BlockingQueue<String>    lineToSendQueue;
+    protected final ScheduledExecutorService executor;
+
+    public BaseProjectImportOutputLineConsumer(String projectName, int delayBetweenMessages) {
+        this.projectName = projectName;
+        lineToSendQueue = new ArrayBlockingQueue<>(1024);
+        executor = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder().setNameFormat(BaseProjectImportOutputLineConsumer.class.getSimpleName() + "-%d")
+                                          .setDaemon(true)
+                                          .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                                          .build());
+        executor.scheduleAtFixedRate(() -> {
+            String lineToSend = null;
+            while (!lineToSendQueue.isEmpty()) {
+                lineToSend = lineToSendQueue.poll();
+            }
+            if (lineToSend == null) {
+                return;
+            }
+            sendOutputLine(lineToSend);
+        }, 0, delayBetweenMessages, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws IOException {
+        executor.shutdown();
+    }
+
+    @Override
+    public void writeLine(String line) throws IOException {
+        try {
+            lineToSendQueue.put(line);
+        } catch (InterruptedException ignored) {
+            // ignore if interrupted
```

Comment:
I would add at least a log message here, that should slightly improve the logging :-)